File: 01_GOF23/DP_examples/C++/03_Behavioral/Visitor/ConcreteVisitor.cpp
#include "ConcreteVisitor.hpp"
#include "ConcreteElement.hpp"
#include <iostream>

// 访问 Computer
void ComputerPartDisplayVisitor::visit(Computer* computer) {
    std::cout << "Displaying Computer" << std::endl;
}

// 访问 Mouse
void ComputerPartDisplayVisitor::visit(Mouse* mouse) {
    std::cout << "Displaying Mouse" << std::endl;
    mouse->operation();
}

// 访问 Keyboard
void ComputerPartDisplayVisitor::visit(Keyboard* keyboard) {
    std::cout << "Displaying Keyboard" << std::endl;
    keyboard->operation();
}

// 访问 Monitor
void ComputerPartDisplayVisitor::visit(Monitor* monitor) {
    std::cout << "Displaying Monitor" << std::endl;
    monitor->operation();
}

File: 01_GOF23/DP_examples/Lua/03_Behavioral/Visitor/ConcreteVisitor.lua
local IComputerPartVisitor = require("Visitor")

-- 具体访问者：ComputerPartDisplayVisitor
local ComputerPartDisplayVisitor = IComputerPartVisitor:new()

function ComputerPartDisplayVisitor:new(o)
    o = o or IComputerPartVisitor:new(o)
    setmetatable(o, self)
    self.__index = self
    return o
end

function ComputerPartDisplayVisitor:visitComputer(computer)
    print("Displaying Computer")
end

function ComputerPartDisplayVisitor:visitMouse(mouse)
    print("Displaying Mouse")
    mouse:operation()
end

function ComputerPartDisplayVisitor:visitKeyboard(keyboard)
    print("Displaying Keyboard")
    keyboard:operation()
end

function ComputerPartDisplayVisitor:visitMonitor(monitor)
    print("Displaying Monitor")
    monitor:operation()
end

return ComputerPartDisplayVisitor
